### PR TITLE
Add documentation to the `Graph` and `ZigZag` traits

### PR DESCRIPTION
**Pull request by schomatis**
This PR aims at (1) reflecting my current understanding of these modules to provide a useful context for future (concrete) questions I might have of the code and (2) leaving some documentation behind that may be useful for the next one reading this code for the first time.

Towards getting some more context around https://github.com/filecoin-project/rust-proofs/issues/455.

### CODE REVIEW

File: storage-proofs/src/crypto/feistel.rs [hunk 1]
```diff
@@ -30,9 +44,17 @@ pub fn permute(
     while u >= num_elements {
         u = encode(u, keys, precomputed)
     }
+    // Since we are representing `num_elements` using an even number of bits,
+    // that can encode many values above it, so keep repeating the operation
+    // until we land in the permitted range.
+    // TODO: Optimization: Do not repeat the entire `encode`, just one round
+    // at a time (following the `keys` order) until where in the `num_elements`
+    // range.
```
[the remark this answers is not in the document]
Comment (porcuquine) [reply storage-proofs/src/crypto/feistel.rs]: I think what you're proposing here is a change to the construction, and I'm not sure a TODO in the code is the best place to introduce the proposal. Let's move this elsewhere or drop the thread (in the interest of conserving bandwidth, as discussed). Either way, let's remove controversial TODOs from this PR so we can merge the parts which are straightforward documentation.

[the remark this answers is not in the document]
Comment (schomatis) [reply storage-proofs/src/crypto/feistel.rs]: Yes.

**Comment by schomatis** [comment]
@porcuquine Removed TODOs.

**Comment by schomatis** [comment]
Ok, so this second commit, `feistel: remove TODOs`, which basically fixes things done in the first one, is the perfect example of what should *not* be merged into `master`, it should have been squashed before merging, as suggested in https://github.com/filecoin-project/rust-fil-proofs/issues/516.

**Pull request merged**